### PR TITLE
feat(rome_js_analyze): `useExponentiationOperator` rule

### PR DESCRIPTION
**Pull request by kaioduarte**
## Summary
Adds an initial implementation of this ESLint rule [prefer-exponentiation-operator](https://eslint.org/docs/latest/rules/prefer-exponentiation-operator). There are some edge cases to be handled. For realistic usage, it looks fine.

## Test Plan

```bash
cargo test -p rome_js_analyze -- use_exponentiation_operator
```

### CODE REVIEW

**Comment by ematipico** [comment]
@kaioduarte considering that this is an initial implementation, would it be possible to create an issue where you can track works that need to be done for this rule?

**Comment by kaioduarte** [comment]
@ematipico here it is https://github.com/rome/tools/issues/3850

File: crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+                    .ok()?
+                    .as_js_name()?
+                    .value_token()
+                    .ok()?
+                    .token_text_trimmed()
+                    == "pow";
+
+                if has_math && has_pow {
+                    return Some(());
+                }
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs:86]: @xunilrj What's your take: Would it be faster to instead use the semantic model to get the global `Math` symbol and then find all references to it instead of running the rule on each static member expression?

File: crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+        let node = ctx.query();
+
+        match node.callee().ok()?.omit_parentheses() {
+            JsAnyExpression::JsStaticMemberExpression(static_member_expr) => {
+                let has_math = static_member_expr
+                    .object()
+                    .ok()?
+                    .omit_parentheses()
+                    .as_reference_identifier()?
+                    .has_name("Math");
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs:74]: We should check that `Math` isn't an alias of a local variable using the semantic model.

File: crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+                    _ => false,
+                };
+
+                if has_math && has_pow {
+                    return Some(());
+                }
+            }
+            _ => {}
+        }
+
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs:126]: Would it make sense to test that the `pow` call has exactly two arguments?

This could also simplify the `action` a bit if you create a `MathPowCall` struct and use it as the state

```rust
struct MathPowCall {
	base: JsAnyExpression,
	exponent: JsAnyExpression
}
```


This safes you the need to iterate over the arguments in the `action`

File: crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+        let mut mutation = ctx.root().begin();
+        let node = ctx.query();
+
+        if !should_suggest_fix(node)? {
+            return None;
+        }
+
+        let [base, exponent] = node.get_arguments_by_index([0, 1]);
+        let mut base_expr = base?.as_js_any_expression()?.clone().omit_parentheses();
+        let mut exponent_expr = exponent?.as_js_any_expression()?.clone().omit_parentheses();
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs:150]: Is it possible to call `omit_parentheses` first. I believe that should remove the need to call clone

```suggestion
        let mut base_expr = base?.as_js_any_expression()?.omit_parentheses();
        let mut exponent_expr = exponent?.as_js_any_expression()?.omit_parentheses();
```

File: crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+fn does_exponent_need_parens(expr: &JsAnyExpression) -> Option<bool> {
+    Some(expr.precedence().ok()? < OperatorPrecedence::Exponential)
+}
+
+/// Determines whether the given parent node needs parens if used as the exponent in an exponentiation binary expression.
+fn does_parent_expression_need_parens(node: &JsCallExpression) -> Option<JsAnyExpression> {
+    // Skips already parenthesized expressions
+    if has_parenthesized_parent_expression(JsAnyExpression::from(node.clone()))? {
+        return None;
+    }
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs:238]: Isn't it sufficient to test if the direct parent is a parenthesized expression? I recommend handling a parenthesized expression in the `match parent.clone()` as an explicit branch

File: crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+
+    let parent = node.parent::<JsAnyExpression>()?;
+
+    let needs_parentheses = match parent.clone() {
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs:242]: It should be possible to use `&parent` as you aren't consuming the `parent` in any of the branches. 
```suggestion
    let needs_parentheses = match &parent {
```

File: crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+            .args()
+            .iter()
+            .filter_map(|arg| {
+                let binding = arg.ok()?;
+                return binding
+                    .as_js_any_expression()?
+                    .as_js_call_expression()
+                    .cloned();
+            })
+            .any(|arg| &arg == node),
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs:259]: Nit: Does the following work
```suggestion
            .any(|arg| {
                let binding = arg.ok()?;
                binding
                    .as_js_any_expression()?
                    .as_js_call_expression() == Some(&node)
            })
```

File: crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs:271]:
```suggestion
            .any(|arg| {
                let binding = arg.ok()?;
                binding
                    .as_js_any_expression()?
                    .as_js_call_expression() == Some(&node)
            })
```

File: crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+                                .elements()
+                                .into_iter()
+                                .next()?
+                                .as_js_template_chunk_element()?
+                                .syntax()
+                                .text_trimmed()
+                                == "pow"
+                    }
+                    _ => false,
+                };
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs:118]: You can use `member_expr.is_string_constant("pow")` instead

File: crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+    Some(
+        args_count == 2
+            && !arguments.l_paren_token().ok()?.has_leading_comments()
+            && !arguments.l_paren_token().ok()?.has_trailing_comments()
+            && !arguments.r_paren_token().ok()?.has_leading_comments()
+            && !arguments.r_paren_token().ok()?.has_trailing_comments()
+            && arguments
+                .args()
+                .into_iter()
+                .filter_map(|arg| arg.ok())
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs:199]: Nit
```suggestion
                .flatten()
```

File: crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs (hunk 4 above)
Comment (kaioduarte) [reply crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs:150]: We can't because `omit_parentheses` takes ownership of the ref :S

File: crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs (hunk 7 above)
Comment (kaioduarte) [reply crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs:259]: It doesn't work because `any` return should be boolean instead of `Option<bool>`

File: crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs (hunk 3 above)
Comment (kaioduarte) [reply crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs:144]: I think it would be good to emit a diagnostic on every `Math.pow` call, I tried to simplify it a bit on a6799ddbf30cf32e57aab825956d88dc3f0764bb.

File: crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+fn should_suggest_fix(node: &JsCallExpression) -> Option<bool> {
+    let arguments = node.arguments().ok()?;
+    let args_count = arguments.args().len();
+
+    Some(
+        args_count == 2
+            && !arguments.l_paren_token().ok()?.has_leading_comments()
+            && !arguments.l_paren_token().ok()?.has_trailing_comments()
+            && !arguments.r_paren_token().ok()?.has_leading_comments()
+            && !arguments.r_paren_token().ok()?.has_trailing_comments()
```
Comment (kaioduarte) [inline crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs:218]: Is there another way to get the comments around the argument? I did it this way because I couldn't find a way to get them.

In this example, `arg.syntax().has_leading_comments()` doesn't return the `/**/` on both args.
```js
Math.pow(/**/a/**/, /**/b/**/)
```

Comment (ematipico) [reply crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs:218]: That's probably because the comments don't belong to the argument but the parentheses/commas. Try using the playground, the syntax tab will show you where the trivia are stored

Comment (kaioduarte) [reply crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs:218]: Yes, that's what I observed but asked just in case there was another method I don't know 😅 and also because ESLint helpers treats them as part of the argument.

File: crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs (hunk 1 above)
Comment (xunilrj) [reply crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs]: Yes. But we would need to be sure that the context has Math configured as global.

Comment (kaioduarte) [reply crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs]: @xunilrj do we have any rule that does that check?

File: crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs (hunk 2 above)
Comment (kaioduarte) [reply crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs]: Addressed on [3dad256](https://github.com/rome/tools/pull/3848/commits/3dad256385ebc538ab297c6be0b13415c3eb9815).

File: crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs [hunk 12]
```diff
@@ -0,0 +1,334 @@
...
+        })
+    }
+
+    fn get_exponent(&self) -> Option<AnyJsExpression> {
+        Some(if self.does_exponent_need_parens()? {
+            parenthesize_any_js_expression(&self.exponent)
+        } else {
+            self.exponent.clone()
+        })
+    }
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs:81]: Nit: I wouldn't expect a getter to return the `base`/`exponent` as is. How about `make_base` and `make_exponent`?

File: crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs [hunk 13]
```diff
@@ -0,0 +1,334 @@
...
+                } else {
+                    false
+                }
+            }
+            _ => false,
+        };
+
+        if has_math_pow {
+            return Some(());
+        }
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs:158]: Nit: We can remove some of the redundant code 

```suggestion
        let object = match node.callee().ok()?.omit_parentheses() {
            AnyJsExpression::JsStaticMemberExpression(static_member_expr) => {
                if !static_member_expr
                    .member()
                    .ok()?
                    .as_js_name()?
                    .value_token()
                    .ok()?
                    .token_text_trimmed()
                    == "pow" {
                    return None;
                }
                    
                static_member_expr.object()
            }
            
            AnyJsExpression::JsComputedMemberExpression(computed_member_expr) => {
                if !computed_member_expr
                    .member()
                    .ok()?
                    .is_string_constant("pow")) {
                    return None;
              }
              
              computed_member_expr.object()
            }
        };
        
        let reference = object.ok()?.omit_parentheses().as_reference_identifier();
         
         
         // verifies that the Math reference is not a local variable
        let is_math_pow = reference.has_name("Math") && model.binding(&reference).is_none();
        is_math_pow.then_some(())
```

File: crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs [hunk 14]
```diff
@@ -0,0 +1,334 @@
...
+
+        if let Some(class_expr) = extends_clause.parent::<JsClassExpression>() {
+            let class_expr = AnyJsExpression::from(class_expr);
+            if does_expression_need_parens(node, &class_expr)? {
+                return Some((true, Some(class_expr)));
+            }
+        }
+    }
+
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs:270]: Nit
```suggestion
    if let Some(parent) = node.parent::<AnyJsExpression>() {
        if does_expression_need_parens(node, &parent)? {
            return Some((true, Some(parent)));
        }
    } else if let Some(extends_clause) = node.parent::<JsExtendsClause>() {
        if extends_clause.parent::<JsClassDeclaration>().is_some() {
            return Some((true, None));
        }

        if let Some(class_expr) = extends_clause.parent::<JsClassExpression>() {
            let class_expr = AnyJsExpression::from(class_expr);
            if does_expression_need_parens(node, &class_expr)? {
                return Some((true, Some(class_expr)));
            }
        }
    }

```

**Review by MichaReiser — APPROVED** [review]

**Review by leops — APPROVED** [review]

File: crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs (hunk 1 above)
Comment (xunilrj) [reply crates/rome_js_analyze/src/analyzers/nursery/use_exponentiation_operator.rs]: Not exactly. Let us leave the rule as it is, and make this change later as an optimization. Like we did for https://github.com/rome/tools/blob/main/crates/rome_js_analyze/src/semantic_analyzers/correctness/no_arguments.rs
